internal/monitor/otelexporters.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package monitor
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/googlecloudplatform/gcsfuse/v2/cfg"
+	"github.com/googlecloudplatform/gcsfuse/v2/common"
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/logger"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"go.opentelemetry.io/contrib/detectors/gcp"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/prometheus"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/resource"
+	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
+)
+
+// SetupOTelMetricExporters sets up the metrics exporters
+func SetupOTelMetricExporters(ctx context.Context, c *cfg.Config) (shutdownFn common.ShutdownFn) {
+	options := make([]metric.Option, 0)
+	opts, shutdownFn := setupPrometheus(c.Metrics.PrometheusPort)
+	options = append(options, opts...)
+	res, err := getResource(ctx, c)
+	if err != nil {
+		logger.Errorf("Error while fetching resource: %v", err)
+	} else {
+		options = append(options, metric.WithResource(res))
+	}
+	meterProvider := metric.NewMeterProvider(options...)
+	otel.SetMeterProvider(meterProvider)
+	return common.JoinShutdownFunc(meterProvider.Shutdown, shutdownFn)
+}
+
+func setupPrometheus(port int64) ([]metric.Option, common.ShutdownFn) {
+	if port <= 0 {
+		return nil, nil
+	}
+	exporter, err := prometheus.New(prometheus.WithoutUnits(), prometheus.WithoutCounterSuffixes(), prometheus.WithoutScopeInfo(), prometheus.WithoutTargetInfo())
+	if err != nil {
+		logger.Errorf("Error while creating prometheus exporter:%v", err)
+		return nil, nil
+	}
+	shutdownCh := make(chan context.Context)
+	done := make(chan interface{})
+	go serveMetrics(port, shutdownCh, done)
+	return []metric.Option{metric.WithReader(exporter)}, func(ctx context.Context) error {
+		shutdownCh <- ctx
+		close(shutdownCh)
+		<-done
+		close(done)
+		return nil
+	}
+}
+
+func serveMetrics(port int64, shutdownCh <-chan context.Context, done chan<- interface{}) {
+	logger.Infof("Serving metrics at localhost:%d/metrics", port)
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", promhttp.Handler())
+	prometheusServer := &http.Server{
+		Addr:           fmt.Sprintf(":%d", port),
+		Handler:        mux,
+		ReadTimeout:    10 * time.Second,
```

Comment:
How did we come up with these defaults?

Reply:
I think these are just some sane values that were chosen. The same timeouts are currently being used in the OC exporters as well.